completion_aggregator/tasks.py

```diff
@@ -0,0 +1,188 @@
+        """
+        Create an aggregation updater for the given user and course.
+
+        Also takes a modulestore instance.
+        """
+        self.user = user
+        self.course_key = course_key
+        with modulestore.bulk_operations(self.course_key):
+            self.course_block_key = self.init_course_block_key(modulestore, self.course_key)
+            self.course_blocks = self.init_course_blocks(self.user, self.course_block_key)
+        self.aggregators = {
+            aggregator.block_key: aggregator for aggregator in Aggregator.objects.filter(
+                user=self.user,
+                course_key=self.course_key,
+            )
+        }
+        self.block_completions = {completion.block_key: completion for completion in self._get_block_completions()}
+
+    def update(self, changed_blocks=frozenset()):
+        """
+        Update the aggregators for the course.
+
+        Takes a set of completable blocks that have been recently updated to
+        inform how to perform the update.  Currently no optimizations are
+        performed based on this information, but in the future they may help
+        cut down on the amount of work performed.
+        """
+        self.update_for_block(self.course_block_key, changed_blocks)
+
+    def update_for_block(self, block, changed_blocks):
+        """
+        Recursive function to perform updates for a given block.
+
+        Dispatches to an appropriate method given the block's completion_mode.
+        """
+        mode = getattr(XBlock.load_class(block.block_type), 'completion_mode', XBlockCompletionMode.COMPLETABLE)
+        if mode == XBlockCompletionMode.EXCLUDED:
+            return self.update_for_excluded()
+        elif mode == XBlockCompletionMode.COMPLETABLE:
+            return self.update_for_completable(block, changed_blocks)
+        elif mode == XBlockCompletionMode.AGGREGATOR:
+            return self.update_for_aggregator(block, changed_blocks)
+
+    def update_for_aggregator(self, block, changed_blocks):
+        """
+        Calculate the new completion values for an aggregator.
+        """
+        total_earned = 0.0
+        total_possible = 0.0
+        last_modified = OLD_DATETIME
+        for child in self._get_children(block):
+            (earned, possible, modified) = self.update_for_block(child, changed_blocks)
+            total_earned += earned
+            total_possible += possible
+            last_modified = max(last_modified, modified)
+        if self._aggregator_needs_update(block, last_modified):
+            Aggregator.objects.submit_completion(
+                user=self.user,
+                course_key=self.course_key,
+                block_key=block,
+                aggregation_name=block.block_type,
+                earned=total_earned,
+                possible=total_possible,
+                last_modified=last_modified,
+            )
+        return (total_earned, total_possible, last_modified)
```

Comment:
These three methods all return the same tuple data structure, maybe it's worth creating a namedtuple?

Reply:
That's a good idea.